collector/bgp.go

```diff
@@ -8,155 +8,88 @@ import (
 
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
 	"github.com/prometheus/client_golang/prometheus"
 	kingpin "gopkg.in/alecthomas/kingpin.v2"
 )
 
 var (
-	bgpSubsystem         = "bgp"
-	bgpPeerMetricPrefix  = "bgp_peer"
-	bgpL2vpnMetricPrefix = "bgp_l2vpn_evpn"
+	bgpSubsystem = "bgp"
 
+	bgpDescMtx   = sync.Mutex{}
 	bgpDesc      map[string]*prometheus.Desc
 	bgpL2vpnDesc map[string]*prometheus.Desc
 
-	bgpErrors           = []error{}
-	totalBGPErrors      = 0.0
-	bgp6Errors          = []error{}
-	totalBGP6Errors     = 0.0
-	bgpL2VPNErrors      = []error{}
-	totalBGPL2VPNErrors = 0.0
-
 	bgpPeerTypes          = kingpin.Flag("collector.bgp.peer-types", "Enable the frr_bgp_peer_types_up metric (default: disabled).").Default("False").Bool()
 	frrBGPDescKey         = kingpin.Flag("collector.bgp.peer-types.keys", "Select the keys from the JSON formatted BGP peer description of which the values will be used with the frr_bgp_peer_types_up metric. Supports multiple values (default: type).").Default("type").Strings()
 	bgpPeerDescs          = kingpin.Flag("collector.bgp.peer-descriptions", "Add the value of the desc key from the JSON formatted BGP peer description as a label to peer metrics. (default: disabled).").Default("False").Bool()
 	bgpPeerDescsText      = kingpin.Flag("collector.bgp.peer-descriptions.plain-text", "Use the full text field of the BGP peer description instead of the value of the JSON formatted desc key (default: disabled).").Default("False").Bool()
 	bgpAdvertisedPrefixes = kingpin.Flag("collector.bgp.advertised-prefixes", "Enables the frr_exporter_bgp_prefixes_advertised_count_total metric which exports the number of advertised prefixes to a BGP peer. This is an option for older versions of FRR that don't have PfxSent field (default: disabled).").Default("False").Bool()
 )
 
-// BGPCollector collects BGP metrics, implemented as per prometheus.Collector interface.
-type BGPCollector struct{}
-
-// NewBGPCollector returns a BGPCollector struct.
-func NewBGPCollector() *BGPCollector {
-	return &BGPCollector{}
+func init() {
+	registerCollector(bgpSubsystem, enabledByDefault, NewBGPCollector)
+	registerCollector(bgpSubsystem+"6", enabledByDefault, NewBGP6Collector)
+	registerCollector(bgpSubsystem+"l2vpn", enabledByDefault, NewBGPL2VPNCollector)
 }
 
-// Name of the collector. Used to populate flag name.
-func (*BGPCollector) Name() string {
-	return bgpSubsystem
-}
-
-// Help describes the metrics this collector scrapes. Used to populate flag help.
-func (*BGPCollector) Help() string {
-	return "Collect BGP Metrics"
-}
-
-// EnabledByDefault describes whether this collector is enabled by default. Used to populate flag default.
-func (*BGPCollector) EnabledByDefault() bool {
-	return true
-}
-
-// Describe implemented as per the prometheus.Collector interface.
-func (*BGPCollector) Describe(ch chan<- *prometheus.Desc) {
-	for _, desc := range getBgpDesc() {
-		ch <- desc
-	}
+type bgpCollector struct {
+	logger log.Logger
```

Comment:
(1) The global variables `bgpDesc` and `bgpL2vpnDesc` should be moved here. Then (2)

Reply:
Done. I have applied this to all other collectors as well.